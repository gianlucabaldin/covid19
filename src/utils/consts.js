export const DEFAUL_MAX_DATA_SIZE = 30;

export const API_ITALY_HYSTORICAL =
  'https://api.covid19api.com/total/dayone/country/italy';

export const API_ITALY_HYSTORICAL_APIFY =
  'https://api.apify.com/v2/datasets/CUdKmb25Z3HjkoDiN/items?format=json&clean=1';
export const API_ITALY_HYSTORICAL_APIFY_SHORT_URL = 'api.apify.com';

export const API_ITALY_REGION =
  'https://openpuglia.org/api/?q=getdatapccovid-19';

export const OPEN_PUGLIA_API = 'openpuglia.org';
export const COVID_19_API = 'covid19api.com';

export const STATUS = {
  CONFIRMED: 'CONFIRMED',
  RECOVERED: 'RECOVERED',
  DEATHS: 'DEATHS',
  ACTIVES: 'ACTIVES',
  INTENSIVE_THERAPY: 'INTENSIVE_THERAPY',
  TOTAL_HOSPITALIZED: 'TOTAL_HOSPITALIZED',
  TOTAL_POSITIVE: 'TOTAL_POSITIVE',
  TAMPONI: 'TAMPONI',
  TOTAL_CASES: 'TOTAL_CASES',
  DECEASED: 'DECEASED',
  NEW_DAILY_POSITIVE: 'NEW_DAILY_POSITIVE',
  DAILY_DECEASED: 'DAILY_DECEASED',
  DAILY_SWABS: 'DAILY_SWABS',
};
