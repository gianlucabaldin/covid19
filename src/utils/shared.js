/* eslint-disable import/prefer-default-export */

/**
 * Check if the value is a number and more than a thousand, in that case
 * call the toLocalString() witch inserts (eventually) the separator, used
 * in Summary
 * I.E. for italian it inserts the "." ad thousand separator
 * @param {number || string} val the value to insert the separator to if > thousand
 */
export const getLocalizedValue = (val) => {
  if (typeof val === 'number') {
    return Number(val).toLocaleString();
  }
  return val;
};
