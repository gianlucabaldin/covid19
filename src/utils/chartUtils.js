// const DEFAUL_MAX_DATA_SIZE = 20;

import { DEFAUL_MAX_DATA_SIZE } from './consts';

// In order to have a "readable" chart, reduce if too much input values
export const reduceData = (data, maxSize = DEFAUL_MAX_DATA_SIZE) => {
  const { length } = data;
  if (length > maxSize) {
    const y = Math.round(length / maxSize);
    data = data.filter((el, i) => Math.round(i % y) === 0);
  }
  return data;
};
