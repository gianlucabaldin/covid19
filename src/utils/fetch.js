import { mockItalyHistoricalResponseApify } from './mock/mockItalyHistoricalResponse_Apify';
import { mockItalyRegionsResponse } from './mock/mockItalyRegionsResponse';
import { API_ITALY_HYSTORICAL_APIFY, API_ITALY_REGION } from './consts';

export const fetchItalyHistoricalAllApify = async (mock = false) => {
  if (mock) {
    console.log('mock = true');
    return mockItalyHistoricalResponseApify;
  }
  console.log('mock = false');
  return fetch(API_ITALY_HYSTORICAL_APIFY).then((res) => res.json());
};

export const fetchItalyRegion = async (mock = false) => {
  if (mock) {
    return mockItalyRegionsResponse;
  }
  return fetch(API_ITALY_REGION).then((res) => res.json());
};
