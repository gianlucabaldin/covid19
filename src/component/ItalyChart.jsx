import React from 'react';
import '../../node_modules/react-vis/dist/style.css';
import { Container, Box } from '@material-ui/core';
import moment from 'moment';
import SwitchInterval from './SwitchInterval';
import Error from './Error';
import DataProvided from './DataProvided';
import { STATUS, API_ITALY_HYSTORICAL_APIFY_SHORT_URL } from '../utils/consts';
// import SwitchInterval from './SwitchInterval';
import Chart from './Chart';

const ItalyChart = ({
  data,
  error,
  loading,
  checked,
  onToggleSwitch,
  width = 500,
}) => {
  const {
    intensiveTherapy,
    totalHospitalized,
    totalPositive,
    newDailyPositive,
    dailyDeceased,
    dailySwabs,
  } = data;
  return (
    <>
      <Box
        marginTop={1}
        marginBottom={3}
        data-id="italy-chart-box"
        style={{
          display: 'flex',
          justifyContent: 'space-between',
          flexDirection: 'row',
          alignItems: 'center',
        }}
      >
        <DataProvided href={API_ITALY_HYSTORICAL_APIFY_SHORT_URL} />

        <SwitchInterval onToggleSwitch={onToggleSwitch} checked={checked} />
      </Box>

      <Box
        style={{
          display: 'flex',
          // justifyContent: '',
          flexDirection: 'row',
          flexWrap: 'wrap',
        }}
      >
        <Chart
          height={300}
          width={600}
          series={[intensiveTherapy]}
          status={[STATUS.INTENSIVE_THERAPY]}
        />
        <Chart
          height={300}
          width={600}
          series={[totalHospitalized]}
          status={[STATUS.TOTAL_HOSPITALIZED]}
        />
        <Chart
          height={300}
          width={600}
          series={[totalPositive]}
          status={[STATUS.TOTAL_POSITIVE]}
        />
        <Chart
          height={300}
          width={600}
          series={[newDailyPositive]}
          status={[STATUS.NEW_DAILY_POSITIVE]}
        />
        <Chart
          height={300}
          width={600}
          series={[dailyDeceased]}
          status={[STATUS.DAILY_DECEASED]}
        />
        <Chart
          height={300}
          width={600}
          series={[dailySwabs]}
          status={[STATUS.DAILY_SWABS]}
        />
      </Box>
    </>
  );
};

export default ItalyChart;
