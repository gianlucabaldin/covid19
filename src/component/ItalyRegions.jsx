import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import MuiTableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import { Box, Hidden, withStyles } from '@material-ui/core';
import { grey } from '@material-ui/core/colors';
import Error from './Error';
import { OPEN_PUGLIA_API } from '../utils/consts';
import DataProvided from './DataProvided';

const useStyles = makeStyles({
  table: {
    minWidth: 650,
  },
  trHeader: {
    backgroundColor: grey[200],
  },
  th: { fontWeight: 700, padding: 0 },
});

const TableCell = withStyles({
  root: {
    border: '1px solid grey',
  },
})(MuiTableCell);

const ItalyRegions = ({ tableData, width }) => {
  const classes = useStyles();

  /** Legend: 
    1- 'ricoverati con sintomi'       hospitalized with symptoms
    2- 'terapia intensiva'            intensive care
    3- 'totale ospedalizzati'         total hospitalized          --> 1 + 2
    4- 'isolamento domiciliare'       home isolation
    5- 'totale positivi'              total positives             --> 3 + 4 
    6- 'variazione totale positivi'   positives variation
    7- 'nuovi positivi'               new positives
    8- 'dimessi guariti'              discharged healed
    9- deceduti                       deaths
    10- 'totale casi'                 discharged healed           --> 5 + 8 + 9
    11- tamponi                       swabs
  */

  if (!tableData || tableData.length === 0) return <Error />;

  // sort alphabetically by region
  tableData.sort((a, b) => (a.regione > b.regione ? 1 : -1));

  return (
    <>
      {/* <Box
       maxHeight={350}
       // width={width || 800}
       style={{ overflowY: 'scroll', paddingBottom: 10 }}
       data-id="italy-regions-box"
     > */}
      <TableContainer component={Paper}>
        <Table
          className={classes.table}
          size="small"
          aria-label="a dense table"
        >
          <TableHead>
            <TableRow key="header" className={classes.trHeader}>
              <TableCell align="center">Region</TableCell>
              <Hidden lgDown>
                <TableCell align="center">Hospitalized With Symptoms</TableCell>
              </Hidden>
              <TableCell align="center">Intensive Care</TableCell>
              <Hidden lgDown>
                <TableCell align="center">Total Hospitalized</TableCell>
              </Hidden>
              <Hidden mdDown>
                <TableCell align="center">Home Isolation</TableCell>
              </Hidden>
              <TableCell align="center">Total Positives</TableCell>
              <Hidden lgDown>
                <TableCell align="center">Positives Variation</TableCell>
              </Hidden>
              <Hidden lgDown>
                <TableCell align="center">New Positives</TableCell>
              </Hidden>
              <Hidden lgDown>
                <TableCell align="center">Discharged Healed</TableCell>
              </Hidden>
              <TableCell align="center">Deaths</TableCell>
              <TableCell align="center">Total Cases</TableCell>
              <Hidden mdDown>
                <TableCell align="center">Swabs</TableCell>
              </Hidden>
            </TableRow>
          </TableHead>
          <TableBody>
            {tableData &&
              tableData.length > 0 &&
              tableData.map((row) => (
                <TableRow key={row.regione} hover>
                  <TableCell align="center">{row.regione}</TableCell>
                  <Hidden lgDown>
                    <TableCell align="center">
                      {row['ricoverati con sintomi']}
                    </TableCell>
                  </Hidden>
                  <TableCell align="center">
                    {row['terapia intensiva']}
                  </TableCell>
                  <Hidden lgDown>
                    <TableCell align="center">
                      {row['totale ospedalizzati']}
                    </TableCell>
                  </Hidden>
                  <Hidden mdDown>
                    <TableCell align="center">
                      {row['isolamento domiciliare']}
                    </TableCell>
                  </Hidden>
                  <TableCell align="center">{row['totale positivi']}</TableCell>
                  <Hidden lgDown>
                    <TableCell align="center">
                      {row['variazione totale positivi']}
                    </TableCell>
                  </Hidden>
                  <Hidden lgDown>
                    <TableCell align="center">
                      {row['nuovi positivi']}
                    </TableCell>
                  </Hidden>
                  <Hidden lgDown>
                    <TableCell align="center">
                      {row['dimessi guariti']}
                    </TableCell>
                  </Hidden>
                  <TableCell align="center">{row.deceduti}</TableCell>
                  <TableCell align="center">{row['totale casi']}</TableCell>
                  <Hidden mdDown>
                    <TableCell align="center">{row.tamponi}</TableCell>
                  </Hidden>
                </TableRow>
              ))}
          </TableBody>
        </Table>
      </TableContainer>

      <DataProvided date={tableData[0].data} href={OPEN_PUGLIA_API} />
      {/* </Box> */}
    </>
  );
};
export default ItalyRegions;
