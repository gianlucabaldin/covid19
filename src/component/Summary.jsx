import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import MuiTableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import { withStyles, Grid } from '@material-ui/core';
import { grey } from '@material-ui/core/colors';
import { useTranslation } from 'react-i18next';
import Error from './Error';
import { getLocalizedValue } from '../utils/shared';
import Loading from './Loading';

const useStyles = makeStyles((theme) => ({
  trHeader: {
    backgroundColor: grey[200],
  },
  th: {
    [theme.breakpoints.down('sm')]: {
      padding: 4,
      '&:last-child': { paddingRight: 4 },
    },
  },
  td: {
    [theme.breakpoints.down('sm')]: {
      padding: 4,
      '&:last-child': { paddingRight: 4 },
    },
  },
}));

const TableCell = withStyles({
  root: {
    border: '1px solid lightgrey',
    textTransform: 'uppercase',
  },
})(MuiTableCell);

const Summary = ({ data, dataTestId, loading = true, error = false }) => {
  const { t } = useTranslation();
  const classes = useStyles();

  if (error) return <Error />;

  return (
    <Grid item>
      <TableContainer component={Paper}>
        <Table
          data-test-id={`${dataTestId}-container-summary`}
          className={classes.table}
          size="small"
          aria-label="a dense table"
          style={{
            marginTop: '8',
            marginBottom: '8',
          }}
        >
          <TableHead>
            <TableRow key="tr-status" className={classes.trHeader}>
              {data &&
                data.length > 0 &&
                data.map((el) => (
                  <TableCell align="center" className={classes.th} key={el.key}>
                    {t('summary.'.concat(el.key))}
                  </TableCell>
                ))}
            </TableRow>
          </TableHead>
          <TableBody>
            <TableRow key="td-value">
              {loading ? (
                <Loading />
              ) : (
                data &&
                data.length > 0 &&
                data.map((el) => (
                  <TableCell
                    align="center"
                    className={classes.td}
                    data-test-id={'summary-'.concat(el.key)}
                    key={el.key}
                  >
                    {getLocalizedValue(el.value)}
                  </TableCell>
                ))
              )}
            </TableRow>
          </TableBody>
        </Table>
      </TableContainer>
    </Grid>
  );
};

// Summary.defaultProps = {
//   data: {
//     totalCases: 0,
//     deceased: 0,
//     tamponi: 0,
//     error: false,
//   },
// };

export default Summary;
