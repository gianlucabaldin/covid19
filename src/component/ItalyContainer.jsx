/* eslint-disable camelcase */
/* eslint-disable react/jsx-props-no-spreading */
import React, { useEffect, useState } from 'react';
import { Grid, Typography } from '@material-ui/core';
import moment from 'moment';
import { useTranslation } from 'react-i18next';
import ChartContainer from './ChartContainer';
import ItalyRegions from './ItalyRegions';
import { fetchItalyRegion, fetchItalyHistoricalAll } from '../utils/fetch';
import Summary from './Summary';
import { processData } from '../utils/chartUtils';
import { SECTIONS, API_APIFY_SHORT_URL } from '../utils/consts';

const summaryInitialStatus = {
  tamponi: 0,
  totalCases: 0,
  deceased: 0,
  lastUpdate: '',
  dataTestId: SECTIONS.ITALY,
  loading: true,
  error: false,
};

const chartInitialStatus = {
  data: [],
  dataTestId: SECTIONS.ITALY,
  error: false,
  loading: true,
  checked: true,
  href: API_APIFY_SHORT_URL,
};

const tableInitialStatud = {
  data: {},
  loading: true,
};

const ItalyContainer = () => {
  const { t } = useTranslation();

  const [summaryData, setSummaryData] = useState({ ...summaryInitialStatus });
  const [chartData, setChartData] = useState({ ...chartInitialStatus });
  const [tableData, setTableData] = useState(tableInitialStatud);
  const [fetchedDataAll, setFetchedDataAll] = useState();

  const extractData = (res) => {
    const {
      intensiveTherapy,
      totalHospitalized,
      totalPositive,
      newDailyPositive,
      dailyDeceased,
      dailySwabs,
    } = processData(res);

    // fill summary with last day-data extracted from previous fetch
    setSummaryData({
      ...summaryData,
      data: [
        { key: 'total-swabs', value: res[res.length - 1].tamponi },
        { key: 'total-cases', value: res[res.length - 1].totalCases },
        { key: 'total-deceased', value: res[res.length - 1].deceased },
        {
          key: 'last-update',
          value: moment(intensiveTherapy[intensiveTherapy.length - 1].x).format(
            'LLL',
          ),
        },
      ],
      loading: false,
      error: false,
    });
    setChartData({
      ...chartInitialStatus,
      data: [
        {
          'intensive-therapy': intensiveTherapy,
        },
        { 'total-hospitalized': totalHospitalized },
        { 'total-positive': totalPositive },
        { 'new-daily-positive': newDailyPositive },
        { 'daily-deceased': dailyDeceased },
        { 'daily-swabs': dailySwabs },
      ],
      loading: false,
    });
    return res;
  };

  // fetch data from public api or mock (see implementation)
  const fetchData = () => {
    fetchItalyHistoricalAll()
      .then((res) => {
        return extractData(res);
      })
      .then((res) => {
        setFetchedDataAll(res);
      })
      .catch(() => {
        setSummaryData({ loading: false, error: true });
        setChartData({ error: true });
      });

    // fill table with fetched data
    fetchItalyRegion()
      .then((res) => {
        setTableData({ data: res, loading: false });
      })
      .catch((err) => {
        // to fill
        console.log('error', err);
      });
  };
  useEffect(() => {
    fetchData();
  }, []);

  const onToggleSwitch = (checked) => {
    const {
      intensiveTherapy,
      totalHospitalized,
      totalPositive,
      newDailyPositive,
      dailyDeceased,
      dailySwabs,
    } = processData(fetchedDataAll, checked);

    setChartData({
      ...chartInitialStatus,
      data: [
        {
          'intensive-therapy': intensiveTherapy,
        },
        { 'total-hospitalized': totalHospitalized },
        { 'total-positive': totalPositive },
        { 'new-daily-positive': newDailyPositive },
        { 'daily-deceased': dailyDeceased },
        { 'daily-swabs': dailySwabs },
      ],
      loading: false,
      checked,
    });
  };

  return (
    <div data-test-id="italy-container">
      <Grid container justify="center">
        <Summary {...summaryData} />
      </Grid>
      <Grid
        container
        style={{ paddingLeft: 16, paddingRight: 16 }}
        justify="center"
      >
        <Grid item>
          <ChartContainer {...chartData} onToggleSwitch={onToggleSwitch} />
        </Grid>
      </Grid>
      <Grid
        container
        style={{ paddingLeft: 16, paddingRight: 16 }}
        justify="center"
      >
        <Grid item style={{ marginTop: 10 }}>
          <Typography variant="h5" style={{ textTransform: 'capitalize' }}>
            {t('regions-details')}
          </Typography>
        </Grid>
      </Grid>
      <Grid
        container
        style={{ paddingLeft: 16, paddingRight: 16 }}
        justify="center"
      >
        <Grid item style={{ marginTop: 10 }}>
          <ItalyRegions {...tableData} />
        </Grid>
      </Grid>
    </div>
  );
};

// ItalyContainer.defaultProps = {
//   data: {},
//   error: false,
// };

export default ItalyContainer;
