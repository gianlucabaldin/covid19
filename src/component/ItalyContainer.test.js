/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { shallow } from 'enzyme';
import ItalyContainer from './ItalyContainer';
import { fetchItalyHistoricalAllApify } from '../utils/fetch';
import ItalyChart from './ItalyChart';
import ItalyRegions from './ItalyRegions';
import Summary from './Summary';

describe('ItalyContainer', () => {
  let wrapper = shallow(<ItalyContainer />);

  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('contains', () => {
    it('ItalyChart component', () => {
      expect(wrapper.find(ItalyChart).length).toEqual(1);
    });

    it('ItalyRegions component', () => {
      expect(wrapper.find(ItalyRegions).length).toEqual(1);
    });

    it('Summary component', () => {
      expect(wrapper.find(Summary).length).toEqual(1);
    });
  });

  describe('fetches (mock) data', () => {
    let chartData;
    beforeAll(async () => {
      chartData = await fetchItalyHistoricalAllApify(true);
    });

    it('renders with its children ItalyChart', () => {
      wrapper = shallow(
        <ItalyContainer>
          <ItalyChart data={chartData} />
        </ItalyContainer>,
      );
      expect(
        wrapper.find('[data-id="italy-container-chart-box"]').length,
      ).toEqual(1);
    });

    // TODO test ItalyRegion and SUmmary as well, too -> problems
  });
});
