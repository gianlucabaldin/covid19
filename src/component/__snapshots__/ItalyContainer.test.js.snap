// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItalyContainer renders properly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="center"
  >
    <Summary
      data={
        Object {
          "deceased": 0,
          "error": false,
          "tamponi": 0,
          "totalCases": 0,
        }
      }
      deceased={0}
      error={false}
      lastUpdate=""
      tamponi={0}
      totalCases={0}
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="center"
    style={
      Object {
        "paddingLeft": 16,
        "paddingRight": 16,
      }
    }
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <ItalyChart
        checked={true}
        data={
          Object {
            "dailyDeceased": Array [],
            "dailySwabs": Array [],
            "intensiveTherapy": Array [],
            "newDailyPositive": Array [],
            "totalHospitalized": Array [],
            "totalPositive": Array [],
          }
        }
        data-id="italy-container-chart-box"
        error={false}
        loading={true}
        onToggleSwitch={[Function]}
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="center"
    style={
      Object {
        "paddingLeft": 16,
        "paddingRight": 16,
      }
    }
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      style={
        Object {
          "marginTop": 10,
        }
      }
    >
      <WithStyles(ForwardRef(Typography))
        variant="h5"
      >
        Region Details
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="center"
    style={
      Object {
        "paddingLeft": 16,
        "paddingRight": 16,
      }
    }
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      style={
        Object {
          "marginTop": 10,
        }
      }
    >
      <ItalyRegions
        data={Object {}}
        loading={true}
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</Fragment>
`;
